pretyping/reductionops.ml

```diff
@@ -675,6 +685,20 @@ struct
     let (t,def) = Parray.to_array t in
     mkArray(u,t,def,ty)
 
+  (* The [flgs] are forwarded form the calling tactic, which might not be what
+     we want. This means that [Eval hnf let_lazy _ _ (1 + 1) (fun n => n).] is
+     reduced to [1 + 1] and not [2]. *)
```

Comment:
The result of this, i.e. `hnf (let_lazy _ _ (1 + 1) (fun n => n) ===> 1 + 1`, does not seem viable because it breaks intuitions, i.e. that `hnf` returns something in head-normal-form. But it seems like this is more likely to be a bug in the implementation of the primitive than the implementation of the arguments.

Reply:
AFAICT this seems to be just an oversight in the current `hnf` implementation. Unlike normal `Const` nodes, `Const` nodes that contain primitives are simply assumed to return a head normal form term. We should fix that by adding a final call to `redrec` in that branch.

Reply:
Actually, I don't think that's right: we do call `redrec` in that branch.

Reply:
See [that line](https://github.com/rlepigre/coq/blob/33eea9153b110149cf804396918e81b7921af23e/pretyping/reductionops.ml#L812).

Reply:
I think the comment might be misplaced. The `hnf` code goes through `tacred`, doesn't it? I hope I am not confusing these files/functions again.

tacred.v has this here:
```
      | Const (cst, _) when is_primitive env cst ->
         (try
            let args =
              List.map_filter_i (fun i a ->
                  match a with CPrimitives.Kwhnf -> Some i | _ -> None)
                (CPrimitives.kind (Option.get (get_primitive env cst))) in
            let stack = reduce_params allowed_reds env sigma stack args in
            whd_const cst env sigma (applist (x, stack)), []
          with Redelimination -> s')
```
where it simply stops after `whd_const` (which is reduction restricted to the constant `cst`).

Reply:
Ah I see what you mean, OK. I guess I got confused what function we were talking about. If I add an outer call to `red_rec` then we do get a head normal form: `S (0 + 1 + (1 + 1))`. However, I think we have to be careful here, since when there are not enough arguments, the constant will still be there, and the code will loop I think.

Reply:
It might be a little hacky but when can just check the number of arguments here? There's no need to call `whd_const` (and `redrec`) if we don't have enough arguments.

Reply:
I don't think that's good enough: the evaluation of the primitive might fail (i.e., `CredNative.red_prim` returns `None`), in which case we'd end up with the same term.

Reply:
I see. That brings us back to the same problem I always find with Coq reductions: the lack of progress tracking. The usual solution is to insert `EConstr.equal` calls or something silly like that.. :(

Reply:
Implemented the fix.